src/main/ngapp/accio/src/app/editor/mask.directive.ts

```diff
@@ -10,41 +11,132 @@ export class MaskDirective {
-  onClick(e: MouseEvent) {
-    if(!this.disable) {
+   *  Listens for mouse movment over appMask, executes if user's mouse is clicked.
+   *  For each movement, the pixel is added to the paintPixels set and then painted 
+   *    on the canvas.
+   */
+  @HostListener('mousemove', ['$event'])  
+  onMouseMove(e: MouseEvent) {
+    if ((this.tool == MaskTool.paint || this.tool == MaskTool.magicWand) && this.mouseDown) {
       const xCoord = e.offsetX;
       const yCoord = e.offsetY;
 
+      //  User moved mouse, use scribble fill. 
+      this.scribbleFill = true;
+
+      //  Get pixel of user moved over.
+      let pixel = this.magicWandService.coordToDataArrayIndex(
+        Math.floor(xCoord / this.scale),
+        Math.floor(yCoord / this.scale), 
+        this.originalImageData.width);
+
+      this.paintPixels.add(pixel);
+
+      console.log('drawing pixel mousemove');
+      //  Fire event to draw pixel
+      this.newPaintEvent.emit(pixel);
+    }
+  }
+
+ /** 
+
+  *  Sends class @param paintPixels of pixels user painted to magic-wand.service
+  */
+
+ /** 
+  *  Executes once user releases mouse 
+  *  Calls magic-wand.service paint/scribbleFloodFill/floodFill function
+  *    depending on what tool is selected and if the user moved the mouse or not. 
+  *  If normal floodFill, then records the x and y offset, retrieves the tolerance and calls the 
+  *    floodfill algorithm. 
+  *  If the user moved the mouse before picking up their mouse, then floodFill does not execute.
+  *    scribble is called instead. 
+  */ 
+  @HostListener('mouseup', ['$event']) 
+  onMouseUp(e: MouseEvent) {
+      this.mouseDown = false;
+    //  If user has paint selected, call paint to add pixels painted to master.
+    if (this.tool == MaskTool.paint) {
+      this.scribbleFill = false;
+      //  TODO: call to save pixels painted in mask once function implemented
+
+    }
+    //  If user has Magic wand selected and they moved the mouse, call scribbleFlood Fill.
+    else if (this.tool == MaskTool.magicWand && this.scribbleFill) {
+      this.scribbleFill = false;
+      //  TODO: uncomment once scribble flood fill implemented
+      // const maskPixels = this.magicWandService.scribbleFloodFill(
+      //   this.originalImageData,
+      //   Math.floor(this.xCoord / this.scale), 
+      //   Math.floor(this.yCoord / this.scale), 
+      //   this.tolerance, 
+      //   this.paintPixels);
+
+      // this.newMaskEvent.emit(maskPixels);
+    }
+    else if (this.tool == MaskTool.magicWand && !this.scribbleFill) {
       console.log('tolerace in mask.dr ' + this.tolerance);
       
       //  Returns an array indices of each pixel in the mask.
       const maskPixels = this.magicWandService.floodfill(
         this.originalImageData, 
-        Math.floor(xCoord / this.scale), 
-        Math.floor(yCoord / this.scale), 
+        Math.floor(this.xCoord / this.scale), 
+        Math.floor(this.yCoord / this.scale), 
```

Comment:
Nothing that you need to change but just mentioning that you might consider creating a function that will take the mouse x,y coordinates and converts it to pixels. Right now this is a math operation because all you need to do is divide by the scale.  However, once you have pan and zoom functionality it will become more complicated.  So when you add that feature all you would have to do it modify that in one place and all of your code is automatically updated.  Contrast that with you having to search over your code for Math.floor and figure out if you need to replace it.

Again you don't need to change it in this PR.  I'm just mentioning it for future guidance.

Reply:
noted! I can do that for the next pr.